core/arch/riscv64/emit_utils.c

```diff
@@ -656,15 +670,48 @@ void
 append_restore_simd_reg(dcontext_t *dcontext, instrlist_t *ilist, bool absolute)
 {
     opnd_t memopnd;
+    uint vtypei;
 
     /* Floating-point register is not SIMD registers in RISC-V, but to be consistent with
      * other architectures, we handle them here.
      */
     for (int reg = DR_REG_F0; reg <= DR_REG_F31; reg++) {
         memopnd = opnd_create_dcontext_field_via_reg_sz(
-            dcontext, REG_NULL, REG_OFFSET(reg), reg_get_size(reg));
+            dcontext, REG_NULL, FREG_OFFSET(reg), reg_get_size(reg));
         APP(ilist, INSTR_CREATE_fld(dcontext, opnd_create_reg(reg), memopnd));
     }
+
+    if (proc_has_feature(FEATURE_VECTOR)) {
+        /* ma:   mask agnostic
+         * ta:   tail agnostic
+         * sew:  selected element width
+         * lmul: vector register group multiplier
+         *
+         *           ma            ta         sew=8       lmul=8 */
+        vtypei = (0b1 << 7) | (0b1 << 6) | (0b000 << 3) | 0b011;
+        memopnd = opnd_create_dcontext_field_via_reg_sz(dcontext, DR_REG_A1, 0,
+                                                        reg_get_size(DR_REG_VR0));
+        APP(ilist,
+            INSTR_CREATE_addi(dcontext, opnd_create_reg(DR_REG_A1),
+                              opnd_create_reg(REG_DCXT),
+                              opnd_create_immed_int(VREG_OFFSET(DR_REG_VR0), OPSZ_12b)));
+        APP(ilist,
+            INSTR_CREATE_vsetvli(dcontext, opnd_create_reg(DR_REG_A0),
+                                 opnd_create_reg(DR_REG_ZERO),
+                                 opnd_create_immed_uint(vtypei, OPSZ_11b)));
+        for (int reg = DR_REG_VR0; reg <= DR_REG_VR31; reg += 8) {
```

Comment:
+=8 doesn't sound right: don't we need to loop over all vector regs?

Reply:
The `vsetvli` sets the configuration of the VPU for all the following vector instructions before the next `vset*vl*` to element width (sew): 8b; vector register group multiplier (lmul): 8, effective vector length (vl): maximum.

lmul=8 means every vector instruction is going to take 8 vector registers as a whole and operate on that. So we can use 4 vector instructions here to copy all 32 vector registers. Linux kernel uses the same approach, refer to https://elixir.bootlin.com/linux/latest/source/arch/riscv/include/asm/vector.h#L103.

Added a comment in the code to explain it.